### PR TITLE
Implement zoom

### PR DESCRIPTION
**Pull request by EltayebAhmed**
Refactored PlotFigure class into a base_plot_figure_class and a base_qt_plot_figure and a concrete class
Implemented zoom in and zoom out functionality.

**To test:**
Bring up any plotting window (either by plotting a slice in mslice or using the plotting.pyplot library and click the zoom in button. 
Then draw using click and drag draw a rectangle over an area to zoom in to that area. It will add the previous view to a an internal stack
Clicking zoom out will undo the last zoom and remove it from the zoom stack. If the stack is empty zoom out will do nothing.

### CODE REVIEW

File: plotting/plot_window/base_qt_plot_window.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                                   QtGui.QSizePolicy.Expanding)
+        FigureCanvas.updateGeometry(self)
+        self._rect_cid = None
+        self._motion_cid = None
+        self._new_axes_limits = [[None,None],[None,None]] # For zooming [[x_lower,x_upper],[y_lower,y_upper]]
+        # TODO implement zoom history stack
+        self._zoom_history_stack = []
+        self._zoom_rect = None
+
+    def zoom_out(self):
```
Comment (EltayebAhmed) [inline plotting/plot_window/base_qt_plot_window.py]: Is putting the zoom related functions in the canvas ok? I did that because most tools i needed were attributes of the canvas, although it would not be difficult to move this code to the window

File: plotting/plot_window/base_qt_plot_window.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    def _redraw_zoom_rect(self,mouse_event):
+        self._new_axes_limits[0][1] = mouse_event.xdata
+        self._new_axes_limits[1][1] = mouse_event.ydata
+        if self._zoom_rect:
+            self._zoom_rect.remove()
+        x,y = self._new_axes_limits[0][0], self._new_axes_limits[1][0]
+        width, height = map(lambda x: x[1]-x[0], self._new_axes_limits)
+        self._zoom_rect = Rectangle((x, y), width, height, fill=False)
+        self.axes.add_patch(self._zoom_rect)
```
Comment (EltayebAhmed) [inline plotting/plot_window/base_qt_plot_window.py]: Is using matplotlib to draw the zoom rectangle (shown while selecting region to zoom in on) OK? 
it comes with much less code overhead compared to using a qpainter and overriding the paintEvent function

**Comment by EltayebAhmed** [comment]
@martyngigg

File: models/projection/powder/mantid_projection_calculator.py [hunk 3]
```diff
@@ -1,22 +1,15 @@
...
-        return '0,.1,10'
-        #TODO AskInstrument scientist about calculating scattering angles
-        input_workspace = mtd[input_workspace]
-        ei = float(input_workspace.getRun().getLogData('Ei').value)
-        ki = sqrt(ei/2.07)
-        kf = sqrt((ei-0.0)/2.07)
-        Qx=ki*1-cos(minTheta)*kf
\ No newline at end of file
+    def calculate_projection(self, input_workspace, output_workspace, axis1, axis2):
+        import random,string
```
Comment (martyngigg) [inline models/projection/powder/mantid_projection_calculator.py]: Is there a reason this has been modified when the PR is related to implementing zooming? This seems completely unrelated.

File: models/workspacemanager/mantid_workspace_provider.py [hunk 4]
```diff
@@ -19,4 +19,9 @@ def rename_workspace(self, selected_workspace, newName):
         return RenameWorkspace(selected_workspace,newName)
 
     def save_nexus(self, workspace, path):
-        SaveNexus(workspace,path)
\ No newline at end of file
+        SaveNexus(workspace,path)
+
+    @staticmethod
+    def get_workspace_handle(workspace_name):
```
Comment (martyngigg) [inline models/workspacemanager/mantid_workspace_provider.py]: What's the reason that this is a static method when all other methods are instance methods?

File: plotting/plot_window/base_qt_plot_window.py (hunk 1 above)
Comment (martyngigg) [inline plotting/plot_window/base_qt_plot_window.py]: How is this handled by the standard canvas? I would suggest following a similar style.

File: plotting/plot_window/base_qt_plot_window.py (hunk 2 above)
Comment (martyngigg) [inline plotting/plot_window/base_qt_plot_window.py]: That seems fine. As long as the interactive performance is okay then this seems reasonable.

File: plotting/pyplot.py [hunk 5]
```diff
@@ -6,6 +6,7 @@ def draw_if_interactive():
     # We will always draw because mslice might be running without matplotlib interactive
     for figure in FigureManager._figures.values():
         figure.canvas.draw()
+        figure.canvas.draw()
```
Comment (martyngigg) [inline plotting/pyplot.py]: Does there need to be 2 draw calls here?

**Comment by EltayebAhmed** [comment]
@martyngigg Issues resolved

**Comment by EltayebAhmed** [comment]
Moving mouse out of bounds while zooming breaks zooming

File: plotting/plot_window/base_qt_plot_window.py [hunk 6]
```diff
@@ -45,10 +45,16 @@ def _start_draw_zoom_rect(self,mouse_event):
 
 
     def _redraw_zoom_rect(self,mouse_event):
+        cache = [self._new_axes_limits[0][:],self._new_axes_limits[1][:]]
         self._new_axes_limits[0][1] = mouse_event.xdata
         self._new_axes_limits[1][1] = mouse_event.ydata
         rect = zip(*self._new_axes_limits)
-        rect = self.axes.transData.transform(rect)
+        try:
```
Comment (EltayebAhmed) [inline plotting/plot_window/base_qt_plot_window.py]: If user releases mouse while out of bounds will result in strange zooming bounds, but if you bring the mouse back in before releasing it will zoom as expected.
Is that acceptable behaviour?

Comment (martyngigg) [inline plotting/plot_window/base_qt_plot_window.py]: @EltayebAhmed I don't think that is acceptable unless there's absolutely on way around it.

**Comment by martyngigg** [comment]
This seems to work as expected. One problem I have though is that I can't really see the zoom rectangle as on my system it is black on dark blue with the current colour map.

**Comment by EltayebAhmed** [comment]
@martyngigg  Is it poorly visible
or totally invisible?

**Comment by martyngigg** [comment]
@EltayebAhmed As we discussed we will leave this entirely for `matplotlib` to handle.

The branch has also developed a conflict.

**Comment by EltayebAhmed** [comment]
@martyngigg After merging in master and removing the custom zoom implementation this branch now effectively contains a refactor of the plot_figure to a base class and concrete class.
It now candidate for review and merging as is, Should i rename the pull request? or open a new one from the same branch

**Comment by martyngigg** [comment]
@EltayebAhmed Is it intentional that the zooming no longer works?

**Comment by EltayebAhmed** [comment]
@martyngigg  yes this branch is now effectively just a refactor and not an implement zoom branch, Should i open a new pull request?

**Comment by martyngigg** [comment]
Yes, please close this & rename the branch then open a new pull request on that branch.
